src/os/spawn.php

```diff
@@ -0,0 +1,173 @@
+<?hh
+
+namespace HH\Lib\Experimental\OS;
+use namespace HH\Lib\C;
+use namespace HH\Lib\Vec;
+use namespace HH\Lib\_Private\_OS;
+use type HH\Lib\OS\{FileDescriptor, IsNotADirectoryException};
+use function HH\Lib\OS\{close, open};
+
+newtype pid_t as int = int;
+newtype ExitCode as int = int;
+
+<<__Sealed(
+  posix_spawn_file_actions_addopen::class,
+  posix_spawn_file_actions_adddup2::class,
+  posix_spawn_file_actions_addchdir_np::class,
+)>>
+interface PosixSpawnFileActionsSetter {}
+
+final class posix_spawn_file_actions_addopen
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public int $filedes,
+    public string $path,
+    public int $oflag,
+    public int $mode,
+  ) {}
+}
+
+final class posix_spawn_file_actions_adddup2
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public FileDescriptor $filedes,
+    public int $newfiledes,
+  ) {}
+}
+
+final class posix_spawn_file_actions_addchdir_np
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(public string $path) {
+  }
+}
+
+type posix_spawn_file_actions_t = vec<PosixSpawnFileActionsSetter>;
+
+const int POSIX_SPAWN_SETPGROUP = 2;
+const int POSIX_SPAWN_SETSID = 128;
```

Comment:
Are these the same on linux and macos? if so, explicitly say so, otherwise, use `_OS\IS_MACOS`

Alternatively, just use `1` and `2` (or `1 << 1` to make the bitmask aspect clearer) and comment that they are truly arbitrary values that we picked, which do not necessarily match any OS value

Reply:
The exact value is not specified in POSIX: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/spawn.h.html


However, since we are mimicking the POSIX API behavior, we can define our own values. `POSIX_SPAWN_SETPGROUP = 2` and `POSIX_SPAWN_SETSID = 128` are the values from GUI libc:
https://sourceware.org/git/?p=glibc.git;a=blob;f=posix/spawn.h;h=c4a81227b04490b817ed4a5784f2e5a47c4ccbf7;hb=HEAD

On the other hand, Apple's libc does not support `POSIX_SPAWN_SETSID` according to https://opensource.apple.com/source/xnu/xnu-1228.0.2/bsd/sys/spawn.h.auto.html

Reply:
Also note that `POSIX_SPAWN_SETSID` is not standard even though it is defined in both macOS and Linux.

Reply:
Also note that `POSIX_SPAWN_SETSID` is not included in `POSIX.1-2017`, even though it is defined in both macOS's and Linux's implementation and will be part of future standard as per https://www.austingroupbugs.net/view.php?id=1044